LSP Followup

norcalli opened the pull request:
Addressing things brought up in https://github.com/neovim/neovim/issues/11389 and beyond.

TODO:
- [x] reduce markdown previews of code.
- [x] refactor method triggers to `lsp.buf.*` methods
- [x] switch to v:lua and get rid of vim interface.
- [x] get rid of filetype config in favor of something from https://github.com/neovim/nvim-lsp
- [x] switch error messages to something which doesn't require `ENTER` because if an LSP goes crazy, it'll block neovim.
- [x] finish textDocument/rename
- [x] rename `builtin_callbacks` to `default_callbacks`
- [x] resolve callback at time of calling using default_callbacks instead of at client creation
- [x] Make hover/signatureHelp preview focusable so you can mess with it.
- [x] textDocument/references
- [x] Finish textDocument/rangeFormatting and textDocument/format
- [x] vim.lsp.util.set_qflist and vim.lsp.util.set_loclist
- [x] test more complex text edits because deletion is probably buggy.
- [x] test more complex text edits because deletion is probably buggy.
- [ ] Improve `update_tagstack`/location stuff.  from bfredl `update_tagstack seems like something that should be primitive. Maybe using 'tagfunc' option sidesteps the entire tagstack issue?`
- [ ] adjust offset encoding of parameters sent to callbacks.
- [ ] Fix completion
- [ ] take improvements from nvim-lsp
- [ ] check race condition in for_each_buffer_client. Should probably just skip closed clients.
About specific callbacks:
- [ ] For servers that only support full sync: we should throttle full buffer updates in some situations like global change.
- [ ] Simplify capability check in client.request (use a table or something)?
- [ ] Diagnostic code guessing guifg from guibg is oddly specific. Diagnostic code should just use LspDiagnostics{severity} which should be defined :hi default and thus overridable by user by standard means. A mechanism for generating bright/dark colors should be separate from that, and invoked together by user if they want to.

Feature requests:

- [ ] Support LocationLink as well as Location for definition etc. if server supports it (allows defining a peekDefinition mapping that previews the target location in a float)
- [ ] signature should highlight current param (I also would like it aligned to function start in code like vim-jedi does, but the info is not directly available from server)
- [ ] Support multiple Location[Link]s (by putting them in the loclist?)
- [ ] Allow overriding the symbolKind and completionItemKind mappings in server configs (if you prefer procedure to function, say; it's particularly useful for "less typical" servers like texlab)
- [ ] Provide a utility function to check if configured servers are running that can be used in the status line/airline/lightline to indicate if server is up or crashed.
- [ ] Support filterText for completion (to only show items matching the already typed text, either as prefix or as substring)


Related issues:
- [ ] improve interface for special types in Lua like an empty dict, because may cause problems for people (and me).
- [ ] add bitop library to core
- [ ] add native lua json functions?

Commit:
Reduce code blocks in markdown previews.

If the preview is just a code block, then use the language as the
filetype instead of using markdown.
This reduces the size of the preview.

Commit:
Bugfix for floating_preview

Don't modify your inputs.

Commit:
Bugfixes.

- Return after an error in RPC.
- Use an empty vim table for serialization.

Commit:
Add vim.uri_to_bufnr

Commit:
Bugfix. Don't use nvim.lua that doesn't exist :)

justinmk commented:
> switch error messages to something which doesn't require ENTER because if an LSP goes crazy, it'll block neovim.

Maybe just `:redraw` if the error message is longer than some reasonable length. That will skip hit-enter prompt.

Commit:
Add lsp.buf and hover implementation.

Commit:
Add everything to lsp.buf and get rid of autoload.

Commit:
Move everything to buf & default_callbacks

- Rename builtin_callbacks to default_callbacks and slightly change its
semantics:
  - No longer contains the default implementations. Instead, any
  default_callbacks will be used in preference for our .buf methods.
  - Add this to the docs.

Commit:
Change callback resolution to be dynamic.

This allows default_callbacks to be specified after client creation to
be considered. Also it simplifies the code.

Commit:
Change error writer to not be annoying.

Commit:
Spaces not tabs.

Commit:
Fix rename support.

Commit:
Satisfy lualint.

Commit:
Use err_message in default_callbacks

Commit:
Extend list_extend to take start/finish.

Commit:
Add full text_edit implementation.

- Implements textDocument/formatting, textDocument/rangeFormatting,
workspace/applyEdit.
TODO:
- still has edge cases around replacement probably. Only tested with
inserts on the same position.

Commit:
Fix reference in rename.

Commit:
Use the apply_text_edits from util.

Commit:
Remove resolve_bufnr/lualint

escorponox commented:
> switch error messages to something which doesn't require ENTER because if an LSP goes crazy, it'll block neovim.

You could use virtual text or floating windows like coc.nvim

![image](https://user-images.githubusercontent.com/5127194/69324662-2a4c2280-0c49-11ea-8fd9-5d31c044fcc7.png)

bfredl commented:
@escorponox We do this already. I think @norcalli meant errors like if the server returned an error from a request, or a malformatted return value or something.

bfredl commented on runtime/lua/vim/lsp.lua:
```diff
@@ -26,6 +27,11 @@ local lsp = {
 -- TODO improve handling of scratch buffers with LSP attached.
 
+local function err_message(...)
+  nvim_err_writeln(table.concat(vim.tbl_flatten{...}))
```
TODO (for myself): future `nvim_message` function should have "truncate" flag. I e show only the first 'cmdheight' line[s] in message area + some indicator, but store the full message in history.

norcalli commented:
I was a bit overzealous in calling rename and formatting "finished." They work in many cases, but some edits are getting corrupted. I'm going to take a look at how other clients do it later.

Commit:
Merge remote-tracking branch 'origin/master' into lsp-followup

Seirdy commented:
Some documentation and example configs would also be appreciated. If said documentation exists elsewhere, links to those locations should suffice.

norcalli commented:
This is how my new settings look for anyone curious to try it out. we just use the buf for most things:

```lua
local function language_client_setup(_)
  local mappings = {
    ["nK"]    = map_cmd [[lua vim.lsp.buf.hover()]];
    ["ngd"]   = map_cmd [[lua vim.lsp.buf.definition()]];
    ["ngr"]   = map_cmd [[lua vim.lsp.buf.rename()]];
    ["ngD"]   = map_cmd [[lua vim.lsp.util.show_line_diagnostics()]];
    ["n af"]  = map_cmd [[lua vim.lsp.buf.formatting()]];
    ["x af"]  = map_xmd [[lua vim.lsp.buf.range_formatting()]];
  }
  nvim.bo.omnifunc = "v:lua.vim.lsp.omnifunc"
  nvim_apply_mappings(mappings, { buffer = true; silent = true; })
end
```

I add it on `on_attach`. Formatting is close to being fixed.

Commit:
Updates

- Use correct implementation of text_edits.
- Send indent options to rangeFormatting and formatting.
- Remove references to vim bindings and filetype from lsp.txt
- Add more examples to docs.
- Add before_init to allow changing initialize_params.

Commit:
Fix hovers staying on bufhidden

Commit:
lualint

Commit:
Account for character length in jump position.

Commit:
Fix position params for encoding.

norcalli commented:
@bfredl @justinmk @h-michael alright, for a first pass, it's in a good state right now to be reviewed and merged. There are still things todo, but this fixes a lot of issues that people who are testing this are experiencing and adds some solid features.

So should we pause here and make another followup or should I continue working on it?

Commit:
Remove comments.

norcalli commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -26,11 +26,85 @@ local function remove_prefix(prefix, word)
+    prefix = lines[i_0]:sub(1, A[2])
+  end
+  new_lines = list_extend({}, new_lines)
+  if #suffix > 0 then
```
ordering here is very deliberate, btw

norcalli commented:
And to prove that formatting works, I reformatted `eval.c`. it only took like 3 seconds even with all the extra copying I do in `set_lines` and compiled fine :D

Commit:
Fix encoding translation in other places.

suessflorian commented:
Doing amazing work @norcalli 🙏

Commit:
Improve performance of util.set_lines + bugfix

Also permit character_offset for col past the end of line (useful in
range formatting).

Commit:
Improve the character_offset code.

h-michael commented:
@norcalli 
Does this have anything to do with this PR, but is there any clear reason to make the hover window appear above the cursor?

norcalli commented:
@h-michael I'm pretty sure that code is mostly the same from your PR. So that seems like a better question for yourself? Do you find the position disagreeable?

h-michael commented:
I intended to implement it to display under the cursor unless it protrudes. Okay, let's stop this topic here.

Commit:
Merge remote-tracking branch 'origin/master' into lsp-followup

Commit:
UI tweaks.

- Hide diagnostics on client exit
- Stop insert on popup focus.
- Hide popup on insertchar (for signature_help)

Commit:
Merge branch 'master' into lsp-followup

Commit:
Merge branch 'master' into lsp-followup

Commit:
Add support for textDocument/references.

Add set_qflist and set_loclist.
- Also add locations_to_items, which calculates byte offsets for
character positions in files and avoids unnecessary operations.

Commit:
Bring vim into local scope

norcalli commented on runtime/doc/lsp.txt:
```diff
@@ -508,43 +422,43 @@ vim.lsp.log_levels
+  start_client {
+    ...
+    on_attach = function(client, bufnr)
+      vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
```
TODO me
```suggestion
      vim.bo(bufnr).omnifunc = 'v:lua.vim.lsp.omnifunc'
```
